### DISCUSSION TITLE
SSR Authentication

### DISCUSSION THREAD

**Discussion by melokki** [nuxt/nuxt#16556] [Questions]
Hello 
I am trying to setup a simple authentication using nuxt 3 with `ssr: true`, but I am failing 

What I have so far:
- [x] II can login from the client
- [x] I can call the api
- [x] I can store a cookie with the user token
- [x] I can retrieve the user based on the token

 so far so good, the problem is as follows

I am trying to create a middleware to check if the user has the token set, if it's set get the user to client so I can show him some details

Here is what I have so far:

<details><summary>store/auth/index.ts</summary>

```ts 
//store/auth/index.ts

export const useAuth = defineStore("auth", {
  state: (): AuthState => {
    return {
      user: <User>{},
      loggedIn: false,
    };
  },
  actions: {
    async login(credentials: UserCredentials) {
      await $fetch<UserToken>("/api/auth/login", {
        method: "POST",
        body: { data: credentials },
      });

      this.fetchUser();
    },
    async fetchUser() {
      const { data } = await $fetch<{ data: User }>("/api/auth/me");

      this.user = data;
      this.loggedIn = true;
    },
  },
});
```
</details>

<details><summary>server/api/auth/login.post.ts</summary>

```ts
// server/api/auth/login.post.ts

export default defineEventHandler(async (event): Promise<UserToken> => {
  const { apiUrl } = useRuntimeConfig();
  const credentials = await readBody(event);

  const { data } = await $fetch<{ data: UserToken }>("/auth/login", {
    baseURL: apiUrl,
    method: "POST",
    body: {
      ...credentials,
    },
  });

  const expires = DateTime.fromISO(data.expires_at).toJSDate();

  setCookie(event, "token", data.token, { expires });

  return data;
});
```
</details>

<details><summary>server/api/auth/me.get.ts</summary>

```ts
// server/api/auth/me.get.ts

export default defineEventHandler(async (event): Promise<User> => {
  const { apiUrl } = useRuntimeConfig();
  const token = getCookie(event, "token");

  const data = await $fetch<User>("/auth/me", {
    baseURL: apiUrl,
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return data;
});
```

</details>

Now, this works good when I am in the client app, but as soon as I refresh the page, the user details are lost.
I tried all sorts of different things, nitro plugin, nuxt plugin, server middleware, nuxt middleware, I can't figure it out how to get the user details on the server and make them accessible in client (in the pinia store)

**Comment by thijsw** [comment]
I would suggest placing the logic to check whether a user is authenticated in a global middleware (middleware that will be run for every route), like `middleware/authentication.global.ts`. It will be called server-side and client-side.

**Comment by frankvdp** [comment]
Running into exactly the same issue here: successful login, working fetches. Token stored on an http_only cookie, reload the page, and the cookie is not passed on the the server endpoints. Did you by any chance resolve this issue somehow?

**Reply by melokki** [reply]
I've created a composable named `useToken`, in it I have the following code

```ts
export const useToken = () => {
  return useState("token", () => "");
};
```

in my `default.vue` file I have the following code


```ts
onBeforeMount(async () => {
  const token = useToken().value;
  if (token) {
    auth.init();
  }
})

```

the `auth.init()` method fetched the user from the API endpint

`auth.init()` is a method in my `auth` pinia store
